homeassistant/components/reolink/host.py

```diff
@@ -0,0 +1,168 @@
+        self._hass: HomeAssistant = hass
+
+        self._clientsession: aiohttp.ClientSession | None = None
+        self._unique_id: str | None = None
+
+        cur_protocol = (
+            DEFAULT_PROTOCOL if CONF_PROTOCOL not in options else options[CONF_PROTOCOL]
+        )
+
+        self._api = Host(
+            config[CONF_HOST],
+            config[CONF_USERNAME],
+            config[CONF_PASSWORD],
+            port=config.get(CONF_PORT),
+            use_https=config.get(CONF_USE_HTTPS),
+            protocol=cur_protocol,
+            timeout=DEFAULT_TIMEOUT,
+        )
+
+    @property
+    def unique_id(self):
+        """Create the unique ID, base for all entities."""
+        return self._unique_id
+
+    @property
+    def api(self):
+        """Return the API object."""
+        return self._api
+
+    async def async_init(self) -> bool:
+        """Connect to Reolink host."""
+        self._api.expire_session()
+
+        if not await self._api.get_host_data():
+            return False
+
+        if self._api.mac_address is None:
+            return False
+
+        enable_onvif = None
+        enable_rtmp = None
+        enable_rtsp = None
+
+        if not self._api.onvif_enabled:
+            _LOGGER.debug(
+                "ONVIF is disabled on %s, trying to enable it", self._api.nvr_name
+            )
+            enable_onvif = True
+
+        if not self._api.rtmp_enabled and self._api.protocol == "rtmp":
+            _LOGGER.debug(
+                "RTMP is disabled on %s, trying to enable it", self._api.nvr_name
+            )
+            enable_rtmp = True
+        elif not self._api.rtsp_enabled and self._api.protocol == "rtsp":
+            _LOGGER.debug(
+                "RTSP is disabled on %s, trying to enable it", self._api.nvr_name
+            )
+            enable_rtsp = True
+
+        if enable_onvif or enable_rtmp or enable_rtsp:
+            if not await self._api.set_net_port(
+                enable_onvif=enable_onvif,
+                enable_rtmp=enable_rtmp,
+                enable_rtsp=enable_rtsp,
+            ):
+                if enable_onvif:
+                    _LOGGER.error(
+                        "Unable to switch on ONVIF on %s. You need it to be ON to receive notifications",
+                        self._api.nvr_name,
+                    )
+
+                if enable_rtmp:
+                    _LOGGER.error(
+                        "Unable to switch on RTMP on %s. You need it to be ON",
+                        self._api.nvr_name,
+                    )
+                elif enable_rtsp:
+                    _LOGGER.error(
+                        "Unable to switch on RTSP on %s. You need it to be ON",
+                        self._api.nvr_name,
+                    )
+
+        if self._unique_id is None:
+            self._unique_id = format_mac(self._api.mac_address)
+
+        return True
+
+    async def update_states(self) -> bool:
```

Comment:
Why does this return a boolean? What does that mean?

Reply:
I looked into the library. It seems the boolean represents if there's an error. Shouldn't we react to an error?

Looks like the library is both logging errors and raising errors in some cases. That's usually bad practice. If the library has decided to raise an error, it's better to let the library user decide what to do with that error.

Debug logging is of course always ok.

Reply:
I would recommend to change the library to follow one practice only. Either trap all exceptions and only log errors, or don't log errors at all and just raise library specific exceptions. It's often easier to control the flow by using exceptions than checking booleans.